openrr-remote/src/lib.rs

```diff
@@ -0,0 +1,735 @@
+            Err(e) => {
+                error!("{}", e);
+                arci::gamepad::GamepadEvent::Unknown
+            }
+        }
+    }
+
+    fn stop(&self) {
+        let mut client = self.client.clone();
+        if let Err(e) = block_in_place(client.stop(())) {
+            error!("{}", e);
+        }
+    }
+}
+
+#[tonic::async_trait]
+impl<C> gamepad_server::Gamepad for RemoteGamepadReceiver<C>
+where
+    C: arci::Gamepad + 'static,
+{
+    async fn next_event(
+        &self,
+        _: tonic::Request<()>,
+    ) -> Result<tonic::Response<pb::GamepadEvent>, tonic::Status> {
+        Ok(tonic::Response::new(self.inner.next_event().await.into()))
+    }
+
+    async fn stop(&self, _: tonic::Request<()>) -> Result<tonic::Response<()>, tonic::Status> {
+        self.inner.stop();
+        Ok(tonic::Response::new(()))
+    }
+}
+
+// =============================================================================
+// Messages
+
+impl From<arci::TrajectoryPoint> for pb::TrajectoryPoint {
+    fn from(val: arci::TrajectoryPoint) -> Self {
+        Self {
+            positions: val.positions,
+            velocities: val.velocities.unwrap_or_default(),
+            time_from_start: Some(val.time_from_start.into()),
+        }
+    }
+}
+
+impl From<pb::TrajectoryPoint> for arci::TrajectoryPoint {
+    fn from(val: pb::TrajectoryPoint) -> Self {
+        Self {
+            positions: val.positions,
+            velocities: if val.velocities.is_empty() {
+                None
+            } else {
+                Some(val.velocities)
+            },
+            time_from_start: val.time_from_start.unwrap().try_into().unwrap(),
+        }
+    }
+}
+
+impl From<arci::BaseVelocity> for pb::BaseVelocity {
+    fn from(val: arci::BaseVelocity) -> Self {
+        Self {
+            x: val.x,
+            y: val.y,
+            theta: val.theta,
+        }
+    }
+}
+
+impl From<pb::BaseVelocity> for arci::BaseVelocity {
+    fn from(val: pb::BaseVelocity) -> Self {
+        Self {
+            x: val.x,
+            y: val.y,
+            theta: val.theta,
+        }
+    }
+}
+
+impl From<arci::Isometry2<f64>> for pb::Isometry2 {
+    fn from(val: arci::Isometry2<f64>) -> Self {
+        Self {
+            rotation: Some(pb::UnitComplex {
+                re: val.rotation.re,
+                im: val.rotation.im,
+            }),
+            translation: Some(pb::Translation2 {
+                x: val.translation.x,
+                y: val.translation.y,
+            }),
+        }
+    }
+}
+
+impl From<pb::Isometry2> for arci::Isometry2<f64> {
+    fn from(val: pb::Isometry2) -> Self {
+        let translation = val.translation.unwrap();
+        let rotation = val.rotation.unwrap();
+        Self::from_parts(
+            nalgebra::Translation2::new(translation.x, translation.y),
+            nalgebra::UnitComplex::from_complex(nalgebra::Complex {
+                re: rotation.re,
+                im: rotation.im,
```

Comment:
why you chose complex for rotation?

Reply:
I copied them from openrr-plugin. If there is another good way to decode/encode Isometry2, suggestions are welcome.

Reply:
I thought that just an angle (float) is enough (without much thought).